File: python/tensorrt_model_connect/runtime_config/schemas/text_trace.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Schema for the ``text_trace`` namespace.

Covers the per-step debug tracing of text-generation decodes that used to
live behind the ``TRTMC_TEXT_STEP_TRACE_*`` environment variables. Session-
only — these knobs are purely debug plumbing; no builder path touches them.
"""

from __future__ import annotations

from tensorrt_model_connect.runtime_config import (
    ConfigField,
    Layer,
    Schema,
    register_schema,
)


_SESSION = frozenset({Layer.SESSION_REQUEST, Layer.PLATFORM_PROFILE})


SCHEMA = Schema(
    namespace="text_trace",
    fields=(
        ConfigField(
            name="step_trace_path",
            type_tag="string",
            default="",  # empty → tracing disabled
            allowed_layers=_SESSION,
        ),
        ConfigField(
            name="step_trace_start_pos",
            type_tag="int32",
            default=0,
            allowed_layers=_SESSION,
            validator=lambda v: isinstance(v, int) and v >= 0,
        ),
        # Large enough to act as "unbounded" in practice without spilling the
        # uint32 range; matches the previous INT32_MAX sentinel in the C++.
        ConfigField(
            name="step_trace_end_pos",
            type_tag="int32",
            default=2_000_000_000,
            allowed_layers=_SESSION,
            validator=lambda v: isinstance(v, int) and v >= 0,
        ),
        ConfigField(
            name="step_trace_topk",
            type_tag="int32",
            default=8,
            allowed_layers=_SESSION,
            validator=lambda v: isinstance(v, int) and v >= 1,
        ),
    ),
)


register_schema(SCHEMA)

File: src/runtime/config/schemas/text_trace.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

// Registration for the "text_trace" namespace schema.
//
// Mirrors python/tensorrt_model_connect/runtime_config/schemas/text_trace.py.

#include "trtmc/config/schemas/text_trace.h"

#include <any>
#include <cstdint>
#include <limits>
#include <set>
#include <string>

namespace trtmc::config::schemas {

namespace {
bool is_nonneg_int32(const std::any& v) {
    if (v.type() != typeid(std::int32_t))
        return false;
    return std::any_cast<std::int32_t>(v) >= 0;
}
bool is_positive_int32(const std::any& v) {
    if (v.type() != typeid(std::int32_t))
        return false;
    return std::any_cast<std::int32_t>(v) >= 1;
}
} // namespace

Schema make_text_trace_schema() {
    const std::set<Layer> session = {Layer::SessionRequest, Layer::PlatformProfile};
    return Schema{
        "text_trace",
        {
            ConfigField{"step_trace_path", "string", std::any{std::string{}}, session, nullptr},
            ConfigField{"step_trace_start_pos", "int32", std::any{std::int32_t{0}}, session,
                        is_nonneg_int32},
            ConfigField{"step_trace_end_pos", "int32", std::any{std::int32_t{2'000'000'000}},
                        session, is_nonneg_int32},
            ConfigField{"step_trace_topk", "int32", std::any{std::int32_t{8}}, session,
                        is_positive_int32},
        },
    };
}

REGISTER_CONFIG_SCHEMA_FACTORY_WITH_MANIFEST(register_text_trace_schema, make_text_trace_schema);
} // namespace trtmc::config::schemas
